### DISCUSSION TITLE
howo to run openroad by reading an existing DEF file

### DISCUSSION THREAD

**Discussion by bittnada** [The-OpenROAD-Project/OpenROAD#5065] [Q&A]
Hello
I have a DEF file in which row, track, components, and pins are written.
The position of the cells are on the step of global placement, obtained by using another method.
Now, I want to finish the rest of the procedures by modifying "test/flow.tcl" file.
I just skipped the steps of pin placement and global placement and  added "read_def [my_def_fiel}'", but it did not work.
I will appreciate if anyone let me know how to solve this problem.
Thanks,

**Comment by maliberty** [comment]
What happened when it "did not work"?  Its not much to go on.

**Comment by bittnada** [comment]
Sorry for my delayed response.

I've just conducted a test using gcd_nangate45.tcl, and I've created my own DEF file containing rows, tracks, pins, nets, and components. I performed a simple test by modifying a flow.tcl script to skip buffer insertion, IO placement, Macro Placement, tap cell insertion, power distribution, and global placement, and then adding "read_def".

After running gcd_nangate45.tcl, I received the following output:

====================================================================================
OpenROAD v2.0-13652-g6fc686431 
Features included (+) or not (-): +Charts +GPU +GUI +Python
This program is licensed under the BSD-3 license. See the LICENSE file for details.
Components of this program may be licensed under more restrictive licenses which must be honored.
[INFO ODB-0227] LEF file: Nangate45/Nangate45_tech.lef, created 22 layers, 27 vias
[INFO ODB-0227] LEF file: Nangate45/Nangate45_stdcell.lef, created 135 library cells
[WARNING STA-0441] set_input_delay relative to a clock defined on the same port/pin not allowed.
[INFO IFP-0001] Added 57 rows of 422 site FreePDK45_38x28_10R_NP_162NW_34O.
[INFO ORD-0048] Loading an additional DEF.
Error: Chip already exists
[INFO RSZ-0058] Using max wire length 693um.
[INFO RSZ-0039] Resized 41 instances.
Placement Analysis
---------------------------------
total displacement      17739.5 u
average displacement       49.0 u
max displacement           62.9 u
original HPWL             280.4 u
legalized HPWL           9839.7 u
delta HPWL                 3410 %

worst slack 0.105
worst slack -0.011
tns -0.011
[INFO CTS-0050] Root buffer is BUF_X4.
[INFO CTS-0051] Sink buffer is BUF_X4.
[INFO CTS-0052] The following clock buffers will be used for CTS:
                    BUF_X4
[INFO CTS-0049] Characterization buffer is BUF_X4.
[INFO CTS-0007] Net "clk" found for clock "core_clock".
[WARNING CTS-0041] Net "clk" has 0 sinks. Skipping...
[WARNING CTS-0083] No clock nets have been found.
[INFO CTS-0008] TritonCTS found 0 clock nets.
[WARNING CTS-0082] No valid clock nets in the design.
[INFO RSZ-0058] Using max wire length 693um.
Placement Analysis
---------------------------------
total displacement          0.0 u
average displacement        0.0 u
max displacement            0.0 u
original HPWL            9839.7 u
legalized HPWL           9839.7 u
delta HPWL                    0 %

[INFO RSZ-0094] Found 1 endpoints with setup violations.
[WARNING RSZ-0075] makeBufferedNet failed for driver _393_/ZN
[WARNING RSZ-0075] makeBufferedNet failed for driver _393_/ZN
[WARNING RSZ-0075] makeBufferedNet failed for driver _393_/ZN
[WARNING RSZ-0075] makeBufferedNet failed for driver _393_/ZN
[INFO RSZ-0041] Resized 5 instances.
[INFO RSZ-0043] Swapped pins on 2 instances.
[INFO RSZ-0033] No hold violations found.
worst slack 0.106
worst slack 0.000
tns 0.000
Placement Analysis
---------------------------------
total displacement         91.3 u
average displacement        0.3 u
max displacement            6.4 u
original HPWL            9840.8 u
legalized HPWL           9925.0 u
delta HPWL                    1 %

[INFO DRT-0149] Reading tech and libs.

Units:                2000
Number of layers:     21
Number of macros:     135
Number of vias:       27
Number of viarulegen: 19

Signal 11 received
Stack trace:
 0# 0x0000564758181A44 in /mnt/work/OpenROAD/bin/bin/openroad
 1# 0x00007F4028F54F10 in /lib/x86_64-linux-gnu/libc.so.6
 2# odb::dbTechVia::getTopLayer() in /mnt/work/OpenROAD/bin/bin/openroad
 3# odb::dbWireEncoder::addTechVia(odb::dbTechVia*) in /mnt/work/OpenROAD/bin/bin/openroad
 4# drt::TritonRoute::stackVias(odb::dbBTerm*, int, int, bool) in /mnt/work/OpenROAD/bin/bin/openroad
 5# drt::TritonRoute::processBTermsAboveTopLayer(bool) in /mnt/work/OpenROAD/bin/bin/openroad
 6# drt::TritonRoute::initDesign() in /mnt/work/OpenROAD/bin/bin/openroad
 7# drt::TritonRoute::pinAccess(std::vector<odb::dbInst*, std::allocator<odb::dbInst*> > const&) in /mnt/work/OpenROAD/bin/bin/openroad
 8# pin_access_cmd(char const*, char const*, char const*, int, int) in /mnt/work/OpenROAD/bin/bin/openroad
 9# 0x00005647589E491B in /mnt/work/OpenROAD/bin/bin/openroad
10# TclNRRunCallbacks in /usr/lib/x86_64-linux-gnu/libtcl8.6.so
11# 0x00007F402E79BFBA in /usr/lib/x86_64-linux-gnu/libtcl8.6.so
12# Tcl_EvalEx in /usr/lib/x86_64-linux-gnu/libtcl8.6.so
13# Tcl_Eval in /usr/lib/x86_64-linux-gnu/libtcl8.6.so
14# gui::TclCmdInputWidget::executeCommand(QString const&, bool, bool) in /mnt/work/OpenROAD/bin/bin/openroad
15# gui::CmdInputWidget::handleEnterKeyPress(QKeyEvent*) in /mnt/work/OpenROAD/bin/bin/openroad
16# gui::TclCmdInputWidget::keyPressEvent(QKeyEvent*) in /mnt/work/OpenROAD/bin/bin/openroad
17# QWidget::event(QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
18# QFrame::event(QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
19# QAbstractScrollArea::event(QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
20# QPlainTextEdit::event(QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
21# QApplicationPrivate::notify_helper(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
22# QApplication::notify(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
23# QCoreApplication::notifyInternal2(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
24# 0x00007F402B1357B5 in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
25# QApplicationPrivate::notify_helper(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
26# QApplication::notify(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
27# QCoreApplication::notifyInternal2(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
28# QGuiApplicationPrivate::processKeyEvent(QWindowSystemInterfacePrivate::KeyEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
29# QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
30# QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) in /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
31# 0x00007F4022595260 in /usr/lib/x86_64-linux-gnu/libQt5XcbQpa.so.5
32# g_main_context_dispatch in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
33# 0x00007F402783C770 in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
34# g_main_context_iteration in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
35# QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
36# QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
37# QCoreApplication::exec() in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
38# gui::startGui(int&, char**, Tcl_Interp*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) in /mnt/work/OpenROAD/bin/bin/openroad
39# ord::tclAppInit(Tcl_Interp*) in /mnt/work/OpenROAD/bin/bin/openroad
40# Tcl_MainEx in /usr/lib/x86_64-linux-gnu/libtcl8.6.so
41# main in /mnt/work/OpenROAD/bin/bin/openroad
42# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6
43# _start in /mnt/work/OpenROAD/bin/bin/openroad
================================================================================================

I also examined the source codes and realized that the testing approach I employed did not yield the desired results. However, I am unsure about how to proceed from the step after global placement by reading the DEF file.

Thanks.
